core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java

```diff
@@ -621,19 +621,33 @@ private Metrics() {} // prevent instantiation
   @NotThreadSafe
   public final class StorageChecker implements HeartbeatExecutor {
 
+    private volatile boolean mClosed = false;
+    private final AtomicReference<Thread> mThread = new AtomicReference<>(null);
+
     @Override
     public void heartbeat() {
+      if (mClosed) {
+        return;
+      }
+      mThread.set(Thread.currentThread());
       try {
         mBlockStore.checkStorage();
       } catch (Exception e) {
         LOG.warn("Failed to check storage: {}", e.toString());
         LOG.debug("Exception: ", e);
       }
+      mThread.set(null);
     }
 
     @Override
     public void close() {
-      // Nothing to clean up
```

Comment:
Can you try swapping below two lines in `DefaultBlockWorker::stop()` instead of this logic?

```
// Stop heart-beat executors and clients.
mResourceCloser.close();
// Stop the base. (closes executors.)
super.stop();
```

Reply:
Hmm so I tried this and it worked (the worker shut down in a timely manner)

But I don't quite understand why. `super.stop()` eventually calls `mExecutorService.shutdownNow()` which should interrupt the threads and stop them from executing.

But if the abstract client doesn't have the logic to respond to interrupts that I added here:
https://github.com/Alluxio/alluxio/pull/11432/files#diff-d7dc60da6f1005850b481cc5792912c5R213-R217

I don't understand why the threads are exiting now?